src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -270,137 +236,132 @@ private AsyncProtocolRequest GetOrCreateProtocolRequest(ref AsyncProtocolRequest
+        private Task WriteSingleChunk<TWriteAdapter>(TWriteAdapter writeAdapter, byte[] buffer, int offset, int count)
+            where TWriteAdapter : struct, ISslWriteAdapter
         {
-            if (asyncRequest != null)
+            // Request a write IO slot.
+            Task ioSlot = writeAdapter.LockAsync();
+            if (!ioSlot.IsCompletedSuccessfully)
             {
-                asyncRequest.SetNextRequest(buffer, offset, count, s_resumeAsyncWriteCallback);
+                // Operation is async and has been queued, return.
+                return WaitForWriteIOSlot(writeAdapter, ioSlot, buffer, offset, count);
             }
 
-            // We loop to this method from the callback.
-            // If the last chunk was just completed from async callback (count < 0), we complete user request.
-            if (count >= 0)
-            {
-                do
-                {
-                    if (count == 0 && !SslStreamPal.CanEncryptEmptyMessage)
-                    {
-                        // If it's an empty message and the PAL doesn't support that,
-                        // we're done.
-                        break;
-                    }
+            byte[] rentedBuffer = ArrayPool<byte>.Shared.Rent(count + FrameOverhead);
+            byte[] outBuffer = rentedBuffer;
 
-                    // Request a write IO slot.
-                    if (_sslState.CheckEnqueueWrite(asyncRequest))
-                    {
-                        // Operation is async and has been queued, return.
-                        return;
-                    }
-
-                    int chunkBytes = Math.Min(count, _sslState.MaxDataSize);
-                    byte[] rentedBuffer = ArrayPool<byte>.Shared.Rent(chunkBytes + FrameOverhead);
-                    byte[] outBuffer = rentedBuffer;
-                    int encryptedBytes = 0;
+            SecurityStatusPal status = _sslState.EncryptData(buffer, offset, count, ref outBuffer, out int encryptedBytes);
 
-                    SecurityStatusPal status = _sslState.EncryptData(buffer, offset, chunkBytes, ref outBuffer, out encryptedBytes);
-                    if (status.ErrorCode != SecurityStatusPalErrorCode.OK)
-                    {
-                        // Re-handshake status is not supported.
-                        ProtocolToken message = new ProtocolToken(null, status);
-                        ArrayPool<byte>.Shared.Return(rentedBuffer);
-                        throw new IOException(SR.net_io_encrypt, message.GetException());
-                    }
+            if (status.ErrorCode != SecurityStatusPalErrorCode.OK)
+            {
+                // Re-handshake status is not supported.
+                ArrayPool<byte>.Shared.Return(rentedBuffer);
+                ProtocolToken message = new ProtocolToken(null, status);
+                return Task.FromException(new IOException(SR.net_io_encrypt, message.GetException()));
```

Comment:
Do we need to do anything to "unlock" in a failure case like this, or has that already been taken care of implicitly?  It's a bit jarring to a "LockAsync" call without a corresponding release, but maybe that's just a minor naming issue if it's actually handled under the covers.

Reply:
Yeah it's related to the way the lock is done APM style under the hood in SslState (something  I hope to change in the near future). The call to FinishWrite releases any lock and that is caught in finally blocks higher up. It's just an interlocked based lock so we could call it a second time for clarity if you wish.

Reply:
If it's correct, we can leave it as-is for now, since your change doesn't appear to be making it worse.  It'd be good to re-evaluate this in the near future as more work is done, as you say.